### PR TITLE
add stricter ABI-compatibility checks for punned types

### PR DESCRIPTION
**Pull request by tersec**

### CODE REVIEW

File: beacon_chain/beacon_chain_db_immutable.nim [hunk 1]
```diff
@@ -71,13 +71,21 @@ type
     current_justified_checkpoint*: Checkpoint
     finalized_checkpoint*: Checkpoint
 
-static:
+func getSizeofSig(x: auto, n: int = 0): seq[(string, int, int)] =
```
Comment (arnetheduck) [inline beacon_chain/beacon_chain_db_immutable.nim:74]: name is for.. debugging?

**Review by arnetheduck — APPROVED** [review]

File: beacon_chain/beacon_chain_db_immutable.nim (hunk 1 above)
Comment (tersec) [reply beacon_chain/beacon_chain_db_immutable.nim:74]: It's meant to make the analogy clear as a "recursive sizeof". I'm open to better names; this one isn't terribly catchy. Could reuse the `isomorphic` descriptor here, but it doesn't seem as appropriate.

Comment (arnetheduck) [reply beacon_chain/beacon_chain_db_immutable.nim:74]: I mean the string that gets returned, it looks unused

Comment (tersec) [reply beacon_chain/beacon_chain_db_immutable.nim:74]: No, it's used as part of `doAssert getSizeofSig(T()) == getSizeofSig(U())`.

It returns a sequence of tuples (as it shows) which collectively create a kind of signature of the object and its transitive closure. The string (field name) being the same is part of the two objects being the same.

It's not strictly required for the ABI-compatibility, true, but it's still useful to verify.

**Pull request merged**